### PR TITLE
Parametric input in single CNN model

### PR DESCRIPTION
**Pull request by sahilyadav27**
The previous pull request was about testing a parametric model in itself and working on the suggestions received there. This pull request combines the parametric model into the single CNN model and also adds a config file showing the structure to define the parametric model's architecture (fully connected layers)

### CODE REVIEW

File: ctlearn/default_models/single_cnn.py [hunk 1]
```diff
@@ -8,6 +8,7 @@ def single_cnn_model(data, model_params):
 
     # Load neural network model
     network_input = tf.keras.Input(shape=data.img_shape, name=f"images")
+    network_input_param = tf.keras.Input(shape=data.prm_shape, name=f"parameters")
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:11]: I think that should break if the parameter list is empty. Try to run the default model, i.e. `-d TRN`,  or comment out the parameter list in your config file with this version. Also, try to test the scenario where you only have the parameter list and no images.

File: ctlearn/default_models/single_cnn.py [hunk 2]
```diff
@@ -19,8 +20,10 @@ def single_cnn_model(data, model_params):
                 model.trainable = trainable_backbone
     else:
         sys.path.append(model_params["model_directory"])
-        engine_module = importlib.import_module(model_params["engine"]["module"])
-        engine = getattr(engine_module, model_params["engine"]["function"])
+        engine_module = importlib.import_module(model_params["engine_cnn"]["module"])
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:23]: Unify the naming convention. Here I would propose engine_cnn_module = ...

File: ctlearn/default_models/single_cnn.py [hunk 3]
```diff
@@ -19,8 +20,10 @@ def single_cnn_model(data, model_params):
...
+        engine_module_param = importlib.import_module(model_params["engine_prm"]["module"])
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:24]: And here:
engine_mlp_module_param = importlib.import_module(model_params["engine_mlp"]["module"])
mlp= multilayer precetron, which is the actual engine used

File: ctlearn/default_models/single_cnn.py [hunk 4]
```diff
@@ -42,11 +45,15 @@ def single_cnn_model(data, model_params):
             )(network_input)
 
         engine_output = engine(network_input, params=model_params, name=backbone_name)
+        engine_output_param = engine_param(network_input_param, params=model_params, name=backbone_name)
 
         output = tf.keras.layers.GlobalAveragePooling2D(
             name=backbone_name + "_global_avgpool"
         )(engine_output)
+        output_param = tf.keras.layers.Flatten()(engine_output_param)
+        concat = tf.keras.layers.Concatenate()([output, output_param])
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:54]: Also try to align the naming convention here:
`concat = tf.keras.layers.Concatenate()([output_cnn, output_mlp])`

File: ctlearn/default_models/single_cnn.py (hunk 2 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:23]: okay I will update these names

File: ctlearn/default_models/single_cnn.py [hunk 5]
```diff
@@ -8,6 +8,7 @@ def single_cnn_model(data, model_params):
 
     # Load neural network model
     network_input = tf.keras.Input(shape=data.img_shape, name=f"images")
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:10]: I would propose something like: `network_input_images` or `network_input_img`

File: ctlearn/default_models/single_cnn.py [hunk 6]
```diff
@@ -42,11 +45,15 @@ def single_cnn_model(data, model_params):
             )(network_input)
 
         engine_output = engine(network_input, params=model_params, name=backbone_name)
+        engine_output_param = engine_param(network_input_param, params=model_params, name=backbone_name)
 
         output = tf.keras.layers.GlobalAveragePooling2D(
             name=backbone_name + "_global_avgpool"
         )(engine_output)
+        output_param = tf.keras.layers.Flatten()(engine_output_param)
```
[the remark this answers is not in the document]
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:53]: The dense layer returns (None, 1, 128) output, while the CNN model returns (None, 128), so I had to add the Flatten layer to make them compatible for concatenating them.

File: ctlearn/default_config_files/singleCNN_param.yml [hunk 7]
```diff
@@ -0,0 +1,493 @@
+# Example config file containing all possible options for CTLearn, whether
```
Comment (TjarkMiener) [inline ctlearn/default_config_files/singleCNN_param.yml:1]: It would be better to update the existing example config. You may want to add a `TRN with parameter` model to the default CTLearn models

Comment (sahilyadav27) [reply ctlearn/default_config_files/singleCNN_param.yml:1]: okay makes sense

File: ctlearn/default_models/single_cnn.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (TjarkMiener) [reply ctlearn/default_models/single_cnn.py:53]: I see. It's because the input vector is not flattened.  Shouldn't have any impact when we apply the flatten layer.

File: ctlearn/default_models/single_cnn.py (hunk 1 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:11]: Yes removing the parameter list will cause an issue. We could have a check here if the parameter list is non-empty, and use a flag so that the parametric model lines only execute if required. Does that seem fine? And for the case when there is only a parameter list and no images, we would ideally want a separate parameter model similar to the previous PR I had, right?

File: ctlearn/default_models/single_cnn.py (hunk 5 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: done

File: ctlearn/default_models/single_cnn.py (hunk 3 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: done

File: ctlearn/default_models/single_cnn.py (hunk 4 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: done

File: ctlearn/default_models/single_cnn.py (hunk 1 above)
Comment (TjarkMiener) [reply ctlearn/default_models/single_cnn.py:11]: Your suggestion with a flag sounds good! I think `data.img_shape` and/or `data.prm_shape` should be none if they are not selected. that could even be your flag. Caution in predict mode we normally select also the parameters to store them to disk in the prediction (Data Level 2) file. Therefore your flag for the parameter should also contain a check the mode, i.e. checking if training is selected.

Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:11]: Okay I will keep that in mind and update this. Thanks!

Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: I have added a flag_prm to toggle between the single CNN model with or without parameters. It is working fine in both cases. Do have a look.

File: ctlearn/default_models/single_cnn.py [hunk 8]
```diff
@@ -8,7 +8,10 @@ def single_cnn_model(data, model_params):
 
     # Load neural network model
     network_input_img = tf.keras.Input(shape=data.img_shape, name=f"images")
-    network_input_param = tf.keras.Input(shape=data.prm_shape, name=f"parameters")
+    flag_prm = 0
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:11]: There is no need for this integer. You could use consistent `data.prm_shape != None` instead.

File: ctlearn/default_models/single_cnn.py [hunk 9]
```diff
@@ -8,7 +8,10 @@ def single_cnn_model(data, model_params):
...
+    if data.prm_shape != None:
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:12]: Have you tested the code also in predict mode? I think here you should be also checking if we are in training mode, i.e. `if data.prm_shape != None and data.mode=="train":`

File: ctlearn/default_models/single_cnn.py [hunk 10]
```diff
@@ -8,7 +8,10 @@ def single_cnn_model(data, model_params):
 
     # Load neural network model
     network_input_img = tf.keras.Input(shape=data.img_shape, name=f"images")
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:10]: I'm not sure if we want to support MLP only without the CNNs here or in a separate file. There is no need for an additional file I guess. @nietootein What are your thoughts about this?
I ran a quick check. It is possible to request only the parameter list without the images from dl1dh.  @sahilyadav27 So I guess you can just check here if images are selected or not with `data.img_shape != None`

File: ctlearn/default_models/single_cnn.py [hunk 11]
```diff
@@ -45,15 +50,13 @@ def single_cnn_model(data, model_params):
...
-        
-
-    return singlecnn_model, [network_input_img, network_input_param]
+        if flag_prm == 1:
+            concat = tf.keras.layers.Concatenate()([output_cnn, output_mlp])
+            singlecnn_model = tf.keras.Model(inputs=[network_input_img, network_input_param], outputs = [concat], name=backbone_name)
+            return singlecnn_model, [network_input_img, network_input_param]
+        else:
+            singlecnn_model = tf.keras.Model(network_input_img, output_cnn, name=backbone_name)
+            return singlecnn_model, [network_input_img]
```
Comment (TjarkMiener) [inline ctlearn/default_models/single_cnn.py:62]: It is better to only return once in a function. I'd suggest appending the model inputs/outputs depending on what is selected in the config. We can switch to slack for quicker interaction.

File: ctlearn/default_models/single_cnn.py (hunk 10 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:10]: I thought having a separate MLP model would make sense because we might want to add the MLP model to other models as well like CNN_RNN and ResNet. So it would be better to have a common MLP model independent of the image model rather than having it inside the Single CNN model?

File: ctlearn/default_models/single_cnn.py (hunk 8 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:11]: I thought it would be quicker to check the flag_prm than to check data.prm_shape again in the if statement on line #28

File: ctlearn/default_models/single_cnn.py (hunk 9 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:12]: ah yes I forgot to add that. I will update this

File: ctlearn/default_models/single_cnn.py (hunk 11 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py:62]: Okay I will change this as well.

File: ctlearn/default_models/single_cnn.py (hunk 10 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: This should work now

File: ctlearn/default_models/single_cnn.py (hunk 8 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: Done

File: ctlearn/default_models/single_cnn.py (hunk 9 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: done

File: ctlearn/default_models/single_cnn.py (hunk 11 above)
Comment (sahilyadav27) [reply ctlearn/default_models/single_cnn.py]: done

**Review by TjarkMiener — COMMENTED** [review]
Looks good! Can you please update the default config files? I think you would need to rename engine to engine_cnn to match your changes. At the moment we can wait with a default model for CNN + mlp till we have some solid/best-performing settings for the mlp. However, you can update the existing example config file with the parameter list and mlp settings you have been using for testing, so that new users can see how to use the mlp properly. Other than that everything is fine. We can update the docs separately.
